src/netconf_confirmed_commit.c

```diff
@@ -0,0 +1,938 @@
+            free(path);
+            return;
+        }
+        free(path);
+    }
+}
+
+static void
+ncc_disarm_timer(void)
+{
+    pthread_mutex_lock(&commit_ctx.lock);
+    timer_delete(commit_ctx.timer);
+    commit_ctx.timer = NULL;
+    clean_backup_directory();
+    pthread_mutex_unlock(&commit_ctx.lock);
+}
+
+static void
+ncc_trigger_timer(void)
+{
+    changes_rollback((union sigval)0);
+    ncc_disarm_timer();
+}
+
+/**
+ * @brief check if NP2SRV_DIR exists, try creating it otherwise
+ *
+ * @return SR_ERR_SYS When failed during creation of the directory.
+ * @return SR_ERR_OK When directory exists or was successfully created.
+ */
+static int
+check_dir(void)
+{
+    int rc = SR_ERR_OK;
+    const char *path = NP2SRV_DIR;
+    struct stat buf;
+
+    printf("%s", path);
+    if (stat(NP2SRV_DIR, &buf) == -1) {
+        VRB("Creating directory \"%s\".", NP2SRV_DIR);
+        if (mkdir(NP2SRV_DIR, S_IRWXU) == -1) {
+            ERR("Failed creating directory \"%s\".", NP2SRV_DIR);
+            rc = SR_ERR_SYS;
+        }
+    }
+    return rc;
+}
+
+/**
+ * @brief Backup a module into a file
+ *
+ * @param[in] session Sysrepo session used to get data of the module.
+ * @param[in] module Module to backup.
+ *
+ * @return SR_ERR_LY When printing into the file failed.
+ * @return SR_ERR_NO_MEMORY When memory ran during allocation
+ * @return SR_ERR_OK On success
+ */
+static int
+backup_module(sr_session_ctx_t *session, struct lys_module *module)
+{
+    int rc = SR_ERR_OK;
+    char *path = NULL, *xpath = NULL;
+    struct lyd_node *node = NULL;
+
+    /* LOCK commit_ctk.lock */
+    pthread_mutex_lock(&commit_ctx.lock);
+
+    VRB("Backing up module: %s", module->name);
+
+    if (asprintf(&xpath, "/%s:*", module->name) == -1) {
+        rc = SR_ERR_NO_MEMORY;
+        goto cleanup;
+    }
+
+    if ((rc = sr_get_data(session, xpath, 0, 0, SR_OPER_NO_STATE, &node))) {
+        ERR("Failed getting configuration of running for module \"%s\".", module->name);
+        goto cleanup;
+    }
+
+    if ((rc = check_dir())) {
+        goto cleanup;
+    }
+    if (asprintf(&path, "%s/%s.json", NP2SRV_DIR, module->name) == -1) {
+        rc = SR_ERR_NO_MEMORY;
+        goto cleanup;
+    }
+    if (lyd_print_path(path, node, LYD_JSON, LY_PRINT_SHRINK)) {
+        rc = SR_ERR_LY;
+        goto cleanup;
+    }
+
+cleanup:
+    lyd_free_tree(node);
+    free(path);
+    free(xpath);
+    /* UNLOCK commit_ctk.lock */
+    pthread_mutex_unlock(&commit_ctx.lock);
+    return rc;
+}
+
+/**
+ * @brief Create the timer and set all the options.
+ *
+ * @param[in] timeout Time after which the timer will start the thread.
+ *
+ * @return SR_ERR_SYS When creating the timer fails.
+ * @return SR_ERR_OK When succeeded.
+ */
+static int
+start_timer_thread(uint32_t timeout)
+{
+    struct sigevent sev;
+    struct itimerspec its;
+    timer_t timer_id;
+
+    /* reset the structures */
+    memset(&sev, 0, sizeof sev);
+    memset(&its, 0, sizeof its);
```

Comment:
This is fine but a but better is to initialize the structure directly, saves redundant lines. Such as
```
struct sigevent sev = {0};
```
Standard says that whatever members you do not initialize are set to `0`.